### PR TITLE
fix(Trino): create `PrestoBaseEngineSpec` base class to share common code between Trino and Presto

### PR DESCRIPTION
**Pull request by dungdm93**
### SUMMARY
#20152 Making the `TrinoEngineSpec` derive from the `PrestoEngineSpec` to reusing code. However, underlay driver of [Trino](https://pypi.org/project/trino/) is different from Presto - [PyHive](https://pypi.org/project/PyHive/) which lead to some issues. (See discussion in #20729)

This PR is rework on that PR by create `PrestoBaseEngineSpec` to share common functions between Presto variants.

Fix:
* #20752
* #20428

Suppress:
* #20729

### TESTING INSTRUCTIONS
Unit tests

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by dungdm93** [comment]
cc @villebro, @john-bodley
Could you guys help me list out issues / missing features of Trino when its base class is changed.

**Comment by dungdm93** [comment]
cc @villebro, @john-bodley
Could you guys help me listing issues, or features in Presto u like to have in Trino.

File: superset/db_engine_specs/presto.py [hunk 1]
```diff
@@ -21,27 +21,24 @@
...
-from sqlalchemy.engine.url import URL
 from sqlalchemy.orm import Session
 from sqlalchemy.sql.expression import ColumnClause, Select
 
 from superset import cache_manager, is_feature_enabled
 from superset.common.db_query_status import QueryStatus
 from superset.databases.utils import make_url_safe
-from superset.db_engine_specs.base import BaseEngineSpec, ColumnTypeMapping
+from superset.db_engine_specs.base import ColumnTypeMapping
+from superset.db_engine_specs.presto_base import PrestoBaseEngineSpec
```
Comment (villebro) [inline superset/db_engine_specs/presto.py:41]: I think we could just as well keep `PrestoBaseEngineSpec` in `superset.db_engine_specs.presto`, like we do for `PostgresBaseEngineSpec`, as it kinda keeps the list of modules in `db_engine_specs` shorter. Not strongly opinionated here, but let's at least keep it consistent; if we do decide to break out the bases, then I'd prefer to do it in a follow-up refactor PR and doing the same for other specs, too.

File: superset/db_engine_specs/trino.py [hunk 2]
```diff
@@ -89,32 +113,6 @@ def get_url_for_impersonation(
...
-        database: Database,
-        inspector: Inspector,
-        schema: Optional[str],
-    ) -> List[str]:
-        return BaseEngineSpec.get_view_names(
-            database=database,
-            inspector=inspector,
-            schema=schema,
-        )
-
```
Comment (villebro) [inline superset/db_engine_specs/trino.py:117]: We can also remove `has_implicit_cancel`, as it's no longer needed (it returns the same as `BaseEngineSpec` which isn't overridden in `PrestoBaseEngineSpec`)

**Review by villebro — COMMENTED** [review]
Great work! Tested to work as expected on Trino with a few comments, LMKWYT.

File: superset/db_engine_specs/trino.py (hunk 2 above)
Comment (dungdm93) [reply superset/db_engine_specs/trino.py:117]: `has_implicit_cancel` is removed

File: superset/db_engine_specs/trino.py [hunk 3]
```diff
@@ -17,36 +17,59 @@
...
     except ImportError:
         pass
 
 logger = logging.getLogger(__name__)
 
 
-class TrinoEngineSpec(PrestoEngineSpec):
+class TrinoEngineSpec(PrestoBaseEngineSpec):
     engine = "trino"
-    engine_aliases = {"trinonative"}  # Required for backwards compatibility.
```
Comment (dungdm93) [inline superset/db_engine_specs/trino.py:47]: `trinonative` in `setup.py` already removed since #20152

**Review by villebro — APPROVED** [review]
LGTM assuming CI doesn't turn up any goblins

**Comment by john-bodley** [comment]
@dungdm93 and @villebro the issue I see with this approach is that it introduced several regressions, i.e., the `trino.latest_partition`, `trino.latest_sub_partition` macros no longer work as this functionality is part of the `PrestoEngineSpec` rather than the `PrestoBaseEngineSpec`. Are there plans to either:

1. Add this functionality to the `PrestoBaseEngineSpec`, or
2. Port this functionality to the `TrinoEngineSpec`.

**Comment by villebro** [comment]
@john-bodley the problem is that PyHive and the new Trino connector are diverging at a rapid pace right now, so extending the Trino spec from the old Presto spec is IMO not a good idea (that actually created a few regressions of its own when we did that). I'm fine moving everything that's known to work on both to `PrestoBaseEngineSpec` (I can chip in with adding tests when I have some spare time).

**Comment by john-bodley** [comment]
@villebro understood. I'm 50/50 about whether option 1 or 2 is the preferred solution. Option 1 might be easiest for now (conditional on how coupled the logic is with the DB-API), but long term I sense option 2 is likely preferred.

**Pull request merged**